gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java

```diff
@@ -421,64 +436,49 @@ public void mkdirs(URI path) throws IOException {
-    List<StorageResourceId> itemIds = new ArrayList<>(subdirs.size() * 2 + 1);
-    for (String subdir : subdirs) {
-      itemIds.add(new StorageResourceId(resourceId.getBucketName(), subdir));
-      if (!Strings.isNullOrEmpty(subdir)) {
-        itemIds.add(
-            new StorageResourceId(resourceId.getBucketName(), StringPaths.toFilePath(subdir)));
+    // In case path is a bucket we just attempt to create it without additional checks
+    if (resourceId.isBucket()) {
+      try {
+        gcs.createBucket(resourceId.getBucketName());
+      } catch (IOException e) {
+        if (e instanceof FileAlreadyExistsException
+            || ApiErrorExtractor.INSTANCE.itemAlreadyExists(e)) {
+          // This means that bucket already exist and we do not need to do anything.
+          logger.atFine().withCause(e).log(
+              "mkdirs: %s already exists, ignoring creation failure", resourceId);
+        } else {
+          throw e;
+        }
       }
+      return;
     }
-    // Add the bucket portion.
-    itemIds.add(new StorageResourceId(resourceId.getBucketName()));
-    logger.atFiner().log("mkdirs: going to create dirs with %s paths", itemIds);
 
-    List<GoogleCloudStorageItemInfo> itemInfos = gcs.getItemInfos(itemIds);
+    resourceId = resourceId.toDirectoryId();
 
-    // Each intermediate path must satisfy one of the following conditions:
-    // -- it does not exist or
-    // -- if it exists, it is a directory
-    //
-    // If any of the intermediate paths violates these requirements then
-    // bail out early so that we do not end up with partial set of
-    // created sub-directories. It is possible that the status of intermediate
-    // paths can change after we make this check therefore this is a
-    // good faith effort and not a guarantee.
-    GoogleCloudStorageItemInfo bucketInfo = null;
-    List<StorageResourceId> subdirsToCreate = new ArrayList<>(subdirs.size());
-    for (GoogleCloudStorageItemInfo info : itemInfos) {
-      if (info.isBucket()) {
-        checkState(bucketInfo == null, "bucketInfo should be null");
-        bucketInfo = info;
-      } else if (info.getResourceId().isDirectory() && !info.exists()) {
-        subdirsToCreate.add(info.getResourceId());
-      } else if (!info.getResourceId().isDirectory() && info.exists()) {
-        throw new FileAlreadyExistsException(
-            "Cannot create directories because of existing file: " + info.getResourceId());
-      }
+    // Before creating a top-level directory we need to check if there are no conflicting files
+    // with the same name as any subdirectory
+    if (options.isEnsureNoConflictingItems()) {
+      checkNoFilesConflictingWithDirs(resourceId);
     }
 
-    if (!checkNotNull(bucketInfo, "bucketInfo should not be null").exists()) {
-      gcs.createBucket(bucketInfo.getBucketName());
+    // Create only a top-level directory because subdirectories will be inferred
```

Comment:
I am a little bit confused by the name here, what exactly does top-level here mean?
My understanding is if we are trying to create `gs://foo/bar/zoo/` then top level is `gs://foo`？But this understanding doesn't seem to match with what code is doing?

Reply:
Good point, reworded.